src/adaptors/http2/http2_adaptor.c

```diff
@@ -770,63 +856,102 @@ static bool compose_and_deliver(qdr_http2_connection_t *conn, qdr_http2_stream_d
+                qd_log(http2_adaptor->log_source, QD_LOG_TRACE, "[C%"PRIu64"][S%"PRId32"][L%"PRIu64"] receive_complete = false and has stream_data->body_buffers in compose_and_deliver", conn->conn_id, stream_data->stream_id, stream_data->in_link->identity);
+                bool q2_blocked;
                 if (stream_data->footer_properties) {
-                    qd_message_compose_5(stream_data->message, header_and_props, stream_data->app_properties, stream_data->body, stream_data->footer_properties, receive_complete);
+                	if (!stream_data->entire_footer_arrived) {
+                		qd_compose_free(header_and_props);
+                		return false;
+                	}
+
+                    qd_message_compose_3(stream_data->message, header_and_props, stream_data->app_properties, receive_complete);
+                    qd_message_stream_data_append(stream_data->message, &stream_data->body_buffers, &q2_blocked);
+                    qd_buffer_list_t existing_buffers;
+                    DEQ_INIT(existing_buffers);
+                    qd_compose_take_buffers(stream_data->footer_properties, &existing_buffers);
+                    qd_message_stream_data_footer_append(stream_data->message, &existing_buffers);
                 }
                 else {
-                    qd_message_compose_4(stream_data->message, header_and_props, stream_data->app_properties, stream_data->body, receive_complete);
+                	qd_message_compose_3(stream_data->message, header_and_props, stream_data->app_properties, receive_complete);
+                	qd_message_stream_data_append(stream_data->message, &stream_data->body_buffers, &q2_blocked);
                 }
-                stream_data->body_data_added = true;
+                stream_data->body_data_added_to_msg = true;
+                conn->q2_blocked = conn->q2_blocked || q2_blocked;
+        		if (conn->q2_blocked) {
+        			qd_log(http2_adaptor->protocol_log_source, QD_LOG_TRACE, "[C%"PRIu64"] q2 is blocked on this connection", conn->conn_id);
+        		}
             }
             else {
-
                 if (stream_data->footer_properties) {
+
+                	if (!stream_data->entire_footer_arrived) {
+                		qd_compose_free(header_and_props);
+                		return false;
+                	}
+
                     //
                     // The footer has already arrived but there was no body. Insert an empty body
                     //
-                    stream_data->body = qd_compose(QD_PERFORMATIVE_BODY_DATA, 0);
-                    qd_message_compose_5(stream_data->message, header_and_props, stream_data->app_properties, stream_data->body, stream_data->footer_properties, receive_complete);
+                    qd_message_compose_3(stream_data->message, header_and_props, stream_data->app_properties, receive_complete);
+                    qd_message_stream_data_append(stream_data->message, &stream_data->body_buffers, 0);
```

Comment:
ditto - no check for q2 activation

Reply:
In this specific case the entire footer has arrived but there is no body (DEQ_SIZE(stream_data->body_buffers) == 0)
We are simply calling the qd_message_stream_data_append() to insert an empty body. This will not trigger q2. Then we call qd_message_stream_data_footer_append() to append the footer.

Reply:
Ok.  Personally I've been unpleasantly surprised by Q2 behavior far to many times to trust any assumption about what it's up to.  And I love asserts(), can't get enough of them.

Reply:
I will add the assert, thanks